### PR TITLE
virtio_mode: Add new cases for three virtio mode: modern, transitional and legacy

### PR DESCRIPTION
**Pull request by lijinlijin**
Check device mode are set as expected and device works normally in each mode.

ID: 1598381

Signed-off-by: Li Jin <lijin@redhat.com>

### CODE REVIEW

**Comment by lijinlijin** [comment]
win2012r2 and rhel.7.6 result all pass(serial cases failure are NOT related to this patch): 
(01/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_netkvm.with_modern: PASS (186.73 s)
 (02/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_netkvm.with_transitional: PASS (186.98 s)
 (03/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_netkvm.with_legacy: PASS (186.20 s)
 (04/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_viorng.with_modern: PASS (63.34 s)
 (05/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_viorng.with_transitional: PASS (63.28 s)
 (06/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_viorng.with_legacy: PASS (63.55 s)
 (07/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_viostor.with_modern: PASS (33.35 s)
 (08/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_viostor.with_transitional: PASS (32.45 s)
 (09/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_viostor.with_legacy: PASS (33.69 s)
 (10/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_modern: PASS (33.00 s)
 (11/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_transitional: PASS (33.79 s)
 (12/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_legacy: PASS (33.73 s)
 (13/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_balloon.with_modern: PASS (138.06 s)
 (14/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_balloon.with_transitional: PASS (145.03 s)
 (15/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.x86_64.io-github-autotest-qemu.virtio_mode.with_balloon.with_legacy: PASS (151.25 s)
 (16/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_netkvm.with_modern: PASS (133.05 s)
 (17/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_netkvm.with_transitional: PASS (254.17 s)
 (18/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_netkvm.with_legacy: PASS (165.34 s)
 (19/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_viorng.with_modern: PASS (61.26 s)
 (20/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_viorng.with_transitional: PASS (61.83 s)
 (21/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_viorng.with_legacy: PASS (61.56 s)
 (22/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_viostor.with_modern: PASS (69.97 s)
 (23/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_viostor.with_transitional: PASS (79.22 s)
 (24/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_viostor.with_legacy: PASS (78.26 s)
 (25/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_modern: PASS (291.95 s)
 (26/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_transitional: PASS (291.78 s)
 (27/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_legacy: PASS (79.30 s)
 (28/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_vioserial.with_modern: ERROR (47.09 s)
 (29/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_vioserial.with_transitional: ERROR (43.92 s)
 (30/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_vioserial.with_legacy: ERROR (43.38 s)
 (31/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_balloon.with_modern: PASS (635.16 s)
 (32/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_balloon.with_transitional: PASS (622.50 s)
 (33/33) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.virtio_mode.with_balloon.with_legacy: PASS (417.88 s)

**Comment by lijinlijin** [comment]
@peixiu @kimi1978wy 
Please help to review this patch, thanks

File: qemu/tests/cfg/virtio_mode.cfg [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        fs_type = ext4
+    format_cmd = "yes | mkfs.${fs_type} {0}"
+    show_dev_cmd = "ls {0}"
+    mount_cmd =  "mkdir -p /media && mount -t ${fs_type} {0} /media"
+    show_mount_cmd = "mount | grep {0}"
+    umount_cmd = "umount {0}"
+    testfile_name = "/media/format_disk-test.txt"
+    writefile_cmd = "echo %s > %s"
+    readfile_cmd = "cat %s"
+    iozone_cmd = "WIN_UTILS:\Iozone\iozone.exe -azR -r 64k -n 1G -g 2G -M -i 0 -i 1 -b iozone.xls -f ${disk_letter}:\testfile"
```
Comment (peixiu) [inline qemu/tests/cfg/virtio_mode.cfg]: 1. No disk_letter param.
2. Should need format param if you want to run iozone test on data disk. format disk need first be executed.
3. Better reduce the param '-i 0 -i 1' if added '-a' param. use -i param will not trigger analyses function in iozone_windows, for automation test,  open the analyses function is better~

File: qemu/tests/cfg/virtio_mode.cfg [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                check_rngd_service = "systemctl status rngd"
+                start_rngd_service = "systemctl start rngd"
+                RHEL.6:
+                    check_rngd_service = "service rngd status"
+                    start_rngd_service = "service rngd start"
+        - with_viostor:
+            driver_name = viostor
+            images += " stg"
+            image_name_stg = "images/storage"
+            image_size_stg = 1G
```
Comment (peixiu) [inline qemu/tests/cfg/virtio_mode.cfg]: if iozone command '-g 2G', here image_size_stg better more than 2G.
'-g' means iozone tested max file size.

File: qemu/tests/cfg/virtio_mode.cfg [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            netperf_test_duration = 120
+            netperf_para_sessions = 1
+            test_protocols = TCP_STREAM
+            netperf_server_link_win = "netserver-2.6.0.exe"
+            netperf_client_link_win = "netperf.exe"
+            server_path_win = "c:\\"
+            client_path_win = "c:\\"
+            netperf_client = ${vms}
+            netperf_server = localhost
+            test_protocols = TCP_STREAM
```
Comment (kimi1978wy) [inline qemu/tests/cfg/virtio_mode.cfg]: There are two "test_protocols" in this config

File: qemu/tests/cfg/virtio_mode.cfg [hunk 4]
```diff
@@ -0,0 +1,143 @@
+- virtio_mode:
+    type = virtio_mode
+    cdroms += " virtio winutils"
```
Comment (kimi1978wy) [inline qemu/tests/cfg/virtio_mode.cfg]: No need to add "winutils", it will automatically be in boot cml.

**Comment by peixiu** [comment]
Looks good to me~
Acked-by: Peixiu phou@redhat.com

**Comment by PaulYuuu** [comment]
Tested on ppc platform. And s390 does not have two parameters "disable-modern" and "disable-legacy".
(01/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_netkvm.with_modern: PASS (84.12 s)
(02/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_netkvm.with_transitional: PASS (76.30 s)
(03/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_netkvm.with_legacy: PASS (75.37 s)
(04/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_viorng.with_modern: PASS (74.49 s)
(05/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_viorng.with_transitional: PASS (73.03 s)
(06/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_viorng.with_legacy: PASS (78.60 s)
(07/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_modern: PASS (77.26 s)
(08/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_transitional: PASS (74.72 s)
(09/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_legacy: PASS (71.03 s)
(10/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_balloon.with_modern: PASS (96.49 s)
(11/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_balloon.with_transitional: PASS (78.86 s)
(12/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.6.ppc64le.io-github-autotest-qemu.virtio_mode.with_balloon.with_legacy: PASS (72.08 s)
(01/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_netkvm.with_modern: PASS (64.50 s)
(02/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_netkvm.with_transitional: PASS (61.23 s)
(03/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_netkvm.with_legacy: PASS (59.31 s)
(04/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_viorng.with_modern: PASS (60.83 s)
(05/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_viorng.with_transitional: PASS (60.64 s)
(06/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_viorng.with_legacy: PASS (61.07 s)
(07/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_modern: PASS (60.52 s)
(08/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_transitional: PASS (62.51 s)
(09/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_vioscsi.with_legacy: PASS (59.43 s)
(10/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_balloon.with_modern: PASS (64.33 s)
(11/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_balloon.with_transitional: PASS (63.65 s)
(12/12) Host_RHEL.m8.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.0.ppc64le.io-github-autotest-qemu.virtio_mode.with_balloon.with_legacy: PASS (63.16 s)

**Comment by PaulYuuu** [comment]
Acked-by: Yihuang Yu <yihyu@redhat.com>

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**